File: frontend/src/types/Fight.ts
export interface Fight {
  id: number;
  video_path: string;
  fps: number;
  width: number;
  height: number;
  created_at: string;
  state: string;
  labeled_at: string | null;
  reported_frames: number | null;
  decoded_frames: number | null;
  segmentation_needs_review: boolean;
  segmentation_review_reason: string | null;
  red_fighter_id: number | null;
  blue_fighter_id: number | null;
  red_fighter_name: string | null;
  blue_fighter_name: string | null;
}

export const STATE_PROGRESS: Record<string, number> = {
  validating: 0,
  queued: 0,
  detecting: 10,
  tracking: 35,
  pose: 45,
  corners: 70,
  scoreboard: 78,
  segmenting: 85,
  analyzing: 92,
  completed: 100,
  labeling_in_progress: 96,
  labeling_complete: 100,
  failed: 0,
  invalid: 0,
};

export const STATE_LABELS: Record<string, string> = {
  validating: 'Validating video',
  queued: 'Queued',
  detecting: 'Detecting fighters',
  tracking: 'Tracking fighters',
  pose: 'Analyzing poses',
  corners: 'Identifying corners',
  scoreboard: 'Reading scoreboard',
  segmenting: 'Segmenting rounds',
  analyzing: 'Processing fight',
  completed: 'Completed',
  labeling_in_progress: 'Labeling in progress',
  labeling_complete: 'Labeling complete',
  failed: 'Failed',
  invalid: 'Invalid video',
};

/** States where the pipeline has stopped moving on its own — no more live SSE updates expected. */
export const TERMINAL_STATES = new Set([
  'completed',
  'failed',
  'invalid',
  'labeling_in_progress',
  'labeling_complete',
]);

/** Fight has real analysis data (events/frames/rounds) ready to review in the Player. */
export const isFightViewable = (state: string): boolean =>
  state === 'completed' || state === 'labeling_complete';

/** Fight has fighter_frames/rounds written and is waiting for the user to manually tag events. */
export const isLabelingReady = (state: string): boolean =>
  state === 'labeling_in_progress';

/** Source video failed full-decode validation and was never queued for processing. */
export const isInvalid = (state: string): boolean => state === 'invalid';

/**
 * Segmentation could not corroborate its own round list against the scoreboard,
 * so the rounds are a detection-only guess and should be confirmed by hand.
 * The pipeline sets this at segmentation time — see ai/video_processing/
 * fight_segmentation.py `_review_verdict`.
 */
export const needsRoundReview = (fight: Fight): boolean =>
  fight.segmentation_needs_review === true;

File: backend/app/models/fight.py
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, Column, ForeignKey, Integer, String, Text, TIMESTAMP, text
from pydantic import BaseModel, ConfigDict

from app.utils.db import Base


class Fight(Base):
    __tablename__ = "fights"

    id = Column(Integer, primary_key=True, index=True)
    video_path = Column(String(500), nullable=False, unique=True)
    fps = Column(Integer, nullable=False)
    width = Column(Integer, nullable=False)
    height = Column(Integer, nullable=False)
    created_at = Column(TIMESTAMP, nullable=False, server_default=text("now()"))
    state = Column(String(32), nullable=False, server_default=text("'queued'"))
    pid = Column(Integer, nullable=True)
    labeled_at = Column(TIMESTAMP, nullable=True)
    reported_frames = Column(Integer, nullable=True)
    decoded_frames = Column(Integer, nullable=True)
    # Segmentation's own verdict on its round list — see ai/database.py
    # set_segmentation_review. Written by the pipeline, never by labelling.
    segmentation_needs_review = Column(
        Boolean, nullable=False, server_default=text("false")
    )
    segmentation_review_reason = Column(Text, nullable=True)
    red_fighter_id = Column(Integer, ForeignKey("fighters.id", ondelete="SET NULL"), nullable=True)
    blue_fighter_id = Column(Integer, ForeignKey("fighters.id", ondelete="SET NULL"), nullable=True)


class FightResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    video_path: str
    fps: int
    width: int
    height: int
    created_at: datetime
    state: str
    labeled_at: Optional[datetime]
    reported_frames: Optional[int]
    decoded_frames: Optional[int]
    segmentation_needs_review: bool = False
    segmentation_review_reason: Optional[str] = None
    red_fighter_id: Optional[int]
    blue_fighter_id: Optional[int]
    red_fighter_name: Optional[str] = None
    blue_fighter_name: Optional[str] = None
